infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -128,85 +128,87 @@ private <E, F> void addEventToSessionEventMap(E event, FsmEventType fsmType, lon
     }
 
     private void processSinkTask(LogReplicationSinkEvent event, RemoteSourceLeadershipManager sourceLeadershipManager) {
-        LogReplicationSession session = sourceLeadershipManager.getSession();
-        synchronized (sourceLeadershipManager) {
-            while (!sessionToSinkEventIdMap.get(session).get(0).equals(event.getEventId())) {
+        String sessionName = sourceLeadershipManager.getSessionName();
+        synchronized (sessionToSinkEventIdMap.get(sessionName)) {
+            while (!sessionToSinkEventIdMap.get(sessionName).get(0).equals(event.getEventId())) {
                 try {
-                    sourceLeadershipManager.wait();
+                    sessionToSinkEventIdMap.get(sessionName).wait();
                 } catch (InterruptedException e) {
-                    log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                    log.error("Wait for session {} was interrupted {}", sessionName, e.getMessage());
                 }
             }
         }
 
         sourceLeadershipManager.processEvent(event);
 
-        synchronized (sourceLeadershipManager) {
-            sessionToSinkEventIdMap.get(session).remove(0);
-            sourceLeadershipManager.notifyAll();
+        synchronized (sessionToSinkEventIdMap.get(sessionName)) {
+            sessionToSinkEventIdMap.get(sessionName).remove(0);
+            sessionToSinkEventIdMap.get(sessionName).notifyAll();
         }
     }
 
     private void processRuntimeTask(LogReplicationRuntimeEvent event, CorfuLogReplicationRuntime fsm) {
-        LogReplicationSession session = fsm.getSession();
+        String sessionName = fsm.getSessionName();
         // for a given session, The fsm event should be processed in the order they are queued. This condition ensures
         // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
-        synchronized (fsm) {
-            while (!sessionToRuntimeEventIdMap.get(session).get(0).equals(event.getEventId())) {
-                try {
-                    fsm.wait();
-                } catch (InterruptedException e) {
-                    log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+        synchronized (sessionToRuntimeEventIdMap.get(sessionName)) {
```

Comment:
No this was changed for the case where a session was stopped and restarted.  Since the hashcode for sessions is not consistent, its better to use the sessionName as the lock. Otherwise there can be deadlocks.

Example: say there are 2 threads. `Session1` has lock on '`fsm`'. Thread1 is waiting for an event to be on position 1 for `Session1`. Now session1 is stopped, but its started again, say `Session1'` with `fsm1'`
While thread1 is still waiting, thread2 now waits for an event of `session1'`. 
So both the threads are waiting on different locks at this point.

There are multiple ways to resolve this case...I went with locking on the session name